lib/Dialects/LinalgTransform/Transforms/TransformInterpreter.cpp

```diff
@@ -393,6 +397,56 @@ executeLowerToLLVMOp(ModuleOp module,
   return pm.run(module);
 }
 
+/// Run enabling transformations (LICM and its variants, single-iteration loop
+/// removal, CSE) on the given function.
+static LogicalResult performEnablerTransformations(
+    FuncOp func, TransformOpMapping &operations,
+    linalg::LinalgEnablingOptions options = linalg::LinalgEnablingOptions()) {
+  MLIRContext *ctx = func->getContext();
+  RewritePatternSet patterns(ctx);
+  linalg::populateLinalgTilingCanonicalizationPatterns(patterns);
+  scf::populateSCFForLoopCanonicalizationPatterns(patterns);
+  if (failed(applyPatternsTrackAndFoldGreedily(func, operations, std::move(patterns))))
+    return failure();
+
+  // This assumes LICM never removes operations so we don't need tracking.
+  if (options.licm) {
+    WalkResult result =
+        func->walk([](LoopLikeOpInterface loopLike) -> WalkResult {
+          return moveLoopInvariantCode(loopLike);
+        });
+    if (result.wasInterrupted())
+      return failure();
+  }
+
+  func.walk([](Operation *op) {
+    (void)llvm::TypeSwitch<Operation *, LogicalResult>(op)
+        .Case<AffineForOp, scf::ForOp>(
+            [](auto loop) { return promoteIfSingleIteration(loop); })
+        .Default([](Operation *) { return success(); });
+  });
+
+  if (options.hoistRedundantVectorTransfers)
+    hoistRedundantVectorTransfers(func);
+  if (options.hoistRedundantVectorTransfersOnTensor)
+    hoistRedundantVectorTransfersOnTensor(func);
```

Comment:
Any reason why these are not handled by normal LICM?

Reply:
This requires an analysis of read/write pairs to check if "canBeHoisted" is true.
This is beyond what LICM supports.
Extending behavior with a more general interface is "on-a-per-need-basis" (i.e. not done).